File: pset2/readability/readability.c
/* program to test the grade level readability of a text
We will use one coleman-liau index
formula is: index = 0.0699*L - 0.296*S - 15.8
L = avg nr letters per 100 words
S = avg nr sentences per 100 words
*/
#include <stdbool.h>
#include <string.h>
#include <stdio.h>
#include <cs50.h>
#include <math.h>

//function declarators
string prompt_text();
int letter_counter(string text);
int word_counter(string text, int nr_letters);
int sentence_counter(string text, int nr_letters);
float index_calc(int nr_letters, int nr_words, int nr_sentences);

int main(void)
{

    // prompt user for text
    string sentence;
    sentence = prompt_text();

    int stringlength = strlen(sentence);

    // count nr of letters
    int nr_letters;
    nr_letters = letter_counter(sentence);

    // count nr of words (spaces)
    int nr_words;
    nr_words = word_counter(sentence, stringlength);

    // count nr of sentences (. , ! , ?)
    int nr_sentences;
    nr_sentences = sentence_counter(sentence, stringlength);

    // calculate index
    float index;
    index = index_calc(nr_letters, nr_words, nr_sentences);
    //printf("grade = %f\n", round(index));
    int grade = (int) round(index);
    if (grade >= 16)
    {
        printf("Grade 16+\n");
    }
    else if (grade < 1)
    {
        printf("Before Grade 1\n");
    }
    else
    {
        printf("Grade %d\n", grade);
    }
    // print Grade x where x is grade level, to nearest int
    // 16+ for 16 and higher grade
    // less than 1 "before grade 1"
    return 0;

}

// prompt user for text
string prompt_text()
{
    string text;
    text = get_string("text: ");
    return text;
}

int letter_counter(string text)
{
    int nr_letters = 0;
    for (int i = 0; i < strlen(text); i++)
    {
        if ((text[i] >= 'a' && text[i] <= 'z') || (text[i] >= 'A' && text[i] <= 'Z'))
        {
            nr_letters++;
        }
    }
    return nr_letters;
}

// word counter function
int word_counter(string text, int nr_letters)
{
    int nr_words = 1;

    for (int i = 0; i < nr_letters; i++)
    {
        if (text[i] == ' ')
        {
            nr_words++;
        }
    }
    return nr_words;
}

// sentence counter function
int sentence_counter(string text, int nr_letters)
{
    int nr_sentences = 0;

    for (int i = 0; i < nr_letters; i++)
    {
        if (text[i] == '!' || text[i] == '.' || text[i] == '?')
        {
            nr_sentences++;
        }
    }
    return nr_sentences;
}

float index_calc(int nr_letters, int nr_words, int nr_sentences)
{
    float numLetters = (float) nr_letters;
    float numWords = (float) nr_words;
    float numSens = (float) nr_sentences;
    float index, l, s;

    l = (numLetters / numWords) * 100;
    s = (numSens / numWords) * 100;
    index = 0.0588 * l - 0.296 * s - 15.8;
    return index;
}

File: problemSets/pset6/readability/readability.py
from cs50 import get_string
import re


def main():

    readability = Readability()
    grade = readability.getIndex()

    # grade 16+
    if grade >= 16:
        print("Grade 16+")
    # grade less than 1
    elif grade < 1:
        print("Before Grade 1")
    # else grade
    else:
        print(f"Grade {grade}")


class Readability:

    # constructor prompts user for text
    def __init__(self):
        self.__text = get_string("Text: ")
        self.__letter_count = self.__letterCounter()
        self.__word_count = self.__wordCounter()
        self.__sentence_count = self.__sentenceCounter()
        self.__index = self.__indexCalculator()

    # count number of alphabetic letters in string
    def __letterCounter(self):
        count = 0
        for i in self.__text.lower():
            if i.isalpha():
                count += 1
        return count

    # count number of words in string
    def __wordCounter(self):
        return len(self.__text.strip().split(" "))

    # count number of sentences in string
    def __sentenceCounter(self):
        count = len(re.split(r'[.!?]+', self.__text)) - 1
        return count

    # calculate the index
    def __indexCalculator(self):
        l = (self.__letter_count / self.__word_count) * 100
        s = (self.__sentence_count / self.__word_count) * 100
        return round(0.0588 * l - 0.296 * s - 15.8)

    # getter for letter count
    def getLetterCount(self):
        return self.__letter_count

    # getter for word count
    def getWordCount(self):
        return self.__word_count

    # getter for sentence count
    def getSentenceCount(self):
        return self.__sentence_count

    # getter for index
    def getIndex(self):
        return self.__index


if __name__ == "__main__":
    main()
